### PR TITLE
kubeadm: set pod-infra-container-image for the kubelet

### LINKED ISSUE #1003: kubeadm pulls pause on update and setup ignoring imageRepository: in config file
## What keywords did you search in kubeadm issues before filing this one?
pause
imageRepository

## Versions

**kubeadm version** (use `kubeadm version`): 1.11.0

**Environment**:
- **Kubernetes version** (use `kubectl version`): 1.11
- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Others**:

## What happened?
Setting up the cluster with config file and the following tag:
imageRepository: privateRegistry:5000/google-containers
worked except the pause image is always pulled from k8s.gcr.io/pause, ignoring the setting.
We set up the system by manually pulling pause (1.10 needs the pause-amd64:3.1) and retag it acoordingly.

We upgraded the cluster from 1.10 to 1.11 and had the same issue. The upgrade fails because pause can not be pulled. After pulling manually from privateRegistry:5000/pause:3.1 and retagging it to k8s.gcr.io:pause3.1, the controlplane comes up again.

So kubeadm does not respect the imageRepository, using private or proxy registries, regarding the pause image.

## What you expected to happen?
Using the proper configured image registry

## How to reproduce it (as minimally and precisely as possible)?
Try to set up k8s with kubeadm from private registry, using the imageRepository Tag

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:
The kubelet allows you to set `--pod-infra-container-image`
(also called `PodSandboxImage` in the kubelet config),
which can be a custom location to the "pause" image in the case
of Docker. Other CRIs are not supported.

Set the CLI flag for the Docker case in flags.go using
WriteKubeletDynamicEnvFile().

NOTE: should be rebased after:
https://github.com/kubernetes/kubernetes/pull/66219

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#1003

**Special notes for your reviewer**:
given we are moving to v1beta1 in 1.12 do we need this in master?

**Release note**:

```release-note
kubeadm: when using Docker, automatically set the kubelet flag `--pod-infra-container-image` ("pause" image) if the user has specified `ImageRepository` in the kubeadm config.
```

/hold
/area kubeadm
/kind bug
/assign @luxas
/cc @rosti
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews

### CODE REVIEW

File: cmd/kubeadm/app/cmd/join.go [hunk 1]
```diff
@@ -289,7 +289,7 @@ func (j *Join) Run(out io.Writer) error {
 	}
 
 	// Write env file with flags for the kubelet to use. Also register taints
-	if err := kubeletphase.WriteKubeletDynamicEnvFile(&j.cfg.NodeRegistration, j.cfg.FeatureGates, true, kubeadmconstants.KubeletRunDirectory); err != nil {
+	if err := kubeletphase.WriteKubeletDynamicEnvFile(&j.cfg.NodeRegistration, j.cfg.FeatureGates, "", true, kubeadmconstants.KubeletRunDirectory); err != nil {
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/join.go]: `""` in case of `JoinConfiguration`?

File: cmd/kubeadm/app/phases/kubelet/flags.go [hunk 2]
```diff
@@ -45,11 +46,12 @@ type kubeletFlagsOpts struct {
 
 // WriteKubeletDynamicEnvFile writes a environment file with dynamic flags to the kubelet.
 // Used at "kubeadm init" and "kubeadm join" time.
-func WriteKubeletDynamicEnvFile(nodeRegOpts *kubeadmapi.NodeRegistrationOptions, featureGates map[string]bool, registerTaintsUsingFlags bool, kubeletDir string) error {
+func WriteKubeletDynamicEnvFile(nodeRegOpts *kubeadmapi.NodeRegistrationOptions, featureGates map[string]bool, pauseImage string, registerTaintsUsingFlags bool, kubeletDir string) error {
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/flags.go]: i don't like how `pauseImage` has to be passed here.

Comment (rosti) [reply cmd/kubeadm/app/phases/kubelet/flags.go]: `featureGates` and `pauseImage` should be replaced with `ClusterConfiguration` once #66219 gets merged.

File: cmd/kubeadm/app/images/images.go [hunk 3]
```diff
@@ -64,6 +64,11 @@ func GetEtcdImage(cfg *kubeadmapi.InitConfiguration) string {
 	return GetGenericArchImage(cfg.ImageRepository, constants.Etcd, etcdImageTag)
 }
 
+// GetPauseImage returns the image for the "pause" container
+func GetPauseImage(cfg *kubeadmapi.InitConfiguration) string {
+	return GetGenericImage(cfg.ImageRepository, "pause", "3.1")
```
Comment (rosti) [inline cmd/kubeadm/app/images/images.go]: I would have taken the opportunity to move the version string into a constant too.

File: cmd/kubeadm/app/cmd/join.go (hunk 1 above)
Comment (rosti) [reply cmd/kubeadm/app/cmd/join.go]: This should take `ClusterConfiguration` and then invoke `GetPauseImage` in `WriteKubeletDynamicEnvFile`. But for that to be done, first we need to merge #66219 .
If setup like this, `kubelet join` can try to use the wrong pause image, thus causing this PR to only partially fix the problem (for masters only).

**Review by rosti — COMMENTED** [review]
Looks great, but if I were you, I would hold it until #66219 gets merged (rebase it on that).

File: cmd/kubeadm/app/cmd/join.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/cmd/join.go]: makes sense, waiting on #66219.
if `ClusterConfiguration` contains `ImageRepository` - great.

but the v1beta1 KEP says:
> `v1beta1` API is going to simplify how a Kubernetes administrator can specify custom images by removing `ImageRepository` and `UnifiedControlPlaneInmage` attributes with the the `image` attribute (simpler, more flexible, and consistent with the `image` attribute already used into the `etcd` object)

so if `ImageRepository` is gone from `v1beta1` things like `pause` won't have a custom image repository.

Comment (rosti) [reply cmd/kubeadm/app/cmd/join.go]: I do think, that `ImageRepository` is going to be kept here. Even if it's removed, it's going to be in favour of having separate image fields in the config for all K8s components. This includes the `pause` image. Given the fact, that the pause image is going to be needed on worked nodes too, it's setting is going to be sitting in the `ClusterConfiguration`. Hence, no matter if we have `ImageRepository` or `PauseImage` fields, they are going to sit in `ClusterConfiguration` and using that one is a safe bet for me (as long as #66219 gets merged).

**Comment by karan** [comment]
+1 to this change.

How would auth for the docker registry work? Would `docker login` suffice.

**Comment by neolit123** [comment]
btw, this is stale because it missed the cycle, but also due to this:
https://github.com/kubernetes/kubernetes/pull/66798/files#diff-4e06863a72cb928f527a6d3a8ad4ec5bR164

the ClusterConfiguration has to be downloaded first.

File: cmd/kubeadm/app/phases/kubelet/flags.go [hunk 4]
```diff
@@ -80,6 +82,9 @@ func buildKubeletArgMap(opts kubeletFlagsOpts) map[string]string {
 		} else {
 			kubeletFlags["cgroup-driver"] = driver
 		}
+		if opts.pauseImage != "" {
```
Comment (kennethredler) [inline cmd/kubeadm/app/phases/kubelet/flags.go]: Should this code take into account a value stored in `opts.nodeRegOpts.kubeletExtraArgs["pod-infra-container-image"]`?

Comment (neolit123) [reply cmd/kubeadm/app/phases/kubelet/flags.go]: i need to revisit this PR, but its doing it the other way around at the moment.
whatever image repo the user sets in the config we pass that to the kubelet extra args to form the pod-infra-container-image URL.

Comment (kennethredler) [reply cmd/kubeadm/app/phases/kubelet/flags.go]: I guess my question was could/should it account for both methods? Seems simple enough.

If `pod-infra-container-image` were specified explicitly in `kubeletExtraArgs` it seems that should be chosen over the value calculated from the user-configured image repo.

Comment (neolit123) [reply cmd/kubeadm/app/phases/kubelet/flags.go]: ok, understood.
i'm pretty sure that the value from kubeletExtraArgs (user value) would already override the value we set dynamically:
https://github.com/kubernetes/kubernetes/blob/0dd46c29ae9300f850882603536f08be846a184c/cmd/kubeadm/app/phases/kubelet/flags.go#L65

**Comment by neolit123** [comment]
/close 
@chuckha will take over the implementation for this. :tada:

**Pull request closed without merging**